InfluxDB panic crashes while parsing "-" as Float

on version 0.10.0:

Error Output Line:
influxdb/models/points.go:1418

Likely Source of Error:
influxdb/models/points.go:1374


on version 0.10.0

Possible Solution:
check for non-numeric characters and parse those with len() == 1 as zero or something else.


--
panic: unable to parse number value '-': strconv.ParseFloat: parsing "-": invalid syntax

goroutine 146484 [running]:
github.com/influxdb/influxdb/models.newFieldsFromBinary(0xc381a7a837, 0x7, 0x175c9, 0x4)
        /tmp/tmp.r1Nea5VXEw/src/github.com/influxdb/influxdb/models/points.go:1418 +0x63d
github.com/influxdb/influxdb/models.(*point).unmarshalBinary(0xc320f65710, 0xc2be8602c0)
        /tmp/tmp.r1Nea5VXEw/src/github.com/influxdb/influxdb/models/points.go:1298 +0x3b
github.com/influxdb/influxdb/models.(*point).Fields(0xc320f65710, 0xc2ba0b9dd0)
        /tmp/tmp.r1Nea5VXEw/src/github.com/influxdb/influxdb/models/points.go:1211 +0x41
github.com/influxdb/influxdb/tsdb.(*Shard).validateSeriesAndFields(0xc2b6ab6840, 0xc364fc6000, 0x181c, 0x1e00, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /tmp/tmp.r1Nea5VXEw/src/github.com/influxdb/influxdb/tsdb/shard.go:420 +0xbfc
github.com/influxdb/influxdb/tsdb.(*Shard).WritePoints(0xc2b6ab6840, 0xc364fc6000, 0x181c, 0x1e00, 0x0, 0x0)
        /tmp/tmp.r1Nea5VXEw/src/github.com/influxdb/influxdb/tsdb/shard.go:202 +0xa6
github.com/influxdb/influxdb/tsdb.(*Store).WriteToShard(0xc2080e22c0, 0x5ec40, 0xc364fc6000, 0x181c, 0x1e00, 0x0, 0x0)
        /tmp/tmp.r1Nea5VXEw/src/github.com/influxdb/influxdb/tsdb/store.go:442 +0x14e
github.com/influxdb/influxdb/cluster.func·002(0x5ec40, 0x1, 0xc364fc6000, 0x181c, 0x1e00)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -69,6 +69,7 @@
 - [#5963](https://github.com/influxdata/influxdb/pull/5963): Fix possible deadlock
 - [#4688](https://github.com/influxdata/influxdb/issues/4688): admin UI doesn't display results for some SHOW queries
 - [#6006](https://github.com/influxdata/influxdb/pull/6006): Fix deadlock while running backups
+- [#5965](https://github.com/influxdata/influxdb/issues/5965): InfluxDB panic crashes while parsing "-" as Float
 
 ## v0.10.2 [2016-03-03]
 

--- a/models/points.go
+++ b/models/points.go
@@ -28,6 +28,7 @@ var (
 	}
 
 	ErrPointMustHaveAField = errors.New("point without fields is unsupported")
+	ErrInvalidNumber       = errors.New("invalid number")
 )
 
 // Point defines the values that will be written to the database
@@ -664,12 +665,12 @@ func scanNumber(buf []byte, i int) (int, error) {
 		i++
 		// There must be more characters now, as just '-' is illegal.
 		if i == len(buf) {
-			return i, fmt.Errorf("invalid number")
+			return i, ErrInvalidNumber
 		}
 	}
 
 	// how many decimal points we've see
-	decimals := 0
+	decimal := false
 
 	// indicates the number is float in scientific notation
 	scientific := false
@@ -690,12 +691,11 @@ func scanNumber(buf []byte, i int) (int, error) {
 		}
 
 		if buf[i] == '.' {
-			decimals++
-		}
-
-		// Can't have more than 1 decimal (e.g. 1.1.1 should fail)
-		if decimals > 1 {
-			return i, fmt.Errorf("invalid number")
+			// Can't have more than 1 decimal (e.g. 1.1.1 should fail)
+			if decimal {
+				return i, ErrInvalidNumber
+			}
+			decimal = true
 		}
 
 		// `e` is valid for floats but not as the first char
@@ -713,16 +713,32 @@ func scanNumber(buf []byte, i int) (int, error) {
 
 		// NaN is an unsupported value
 		if i+2 < len(buf) && (buf[i] == 'N' || buf[i] == 'n') {
-			return i, fmt.Errorf("invalid number")
+			return i, ErrInvalidNumber
 		}
 
 		if !isNumeric(buf[i]) {
-			return i, fmt.Errorf("invalid number")
+			return i, ErrInvalidNumber
 		}
 		i++
 	}
-	if isInt && (decimals > 0 || scientific) {
-		return i, fmt.Errorf("invalid number")
+
+	if isInt && (decimal || scientific) {
+		return i, ErrInvalidNumber
+	}
+
+	numericDigits := i - start
+	if isInt {
+		numericDigits--
+	}
+	if decimal {
+		numericDigits--
+	}
+	if buf[start] == '-' {
+		numericDigits--
+	}
+
+	if numericDigits == 0 {
+		return i, ErrInvalidNumber
 	}
 
 	// It's more common that numbers will be within min/max range for their type but we need to prevent
@@ -732,7 +748,7 @@ func scanNumber(buf []byte, i int) (int, error) {
 	if isInt {
 		// Make sure the last char is an 'i' for integers (e.g. 9i10 is not valid)
 		if buf[i-1] != 'i' {
-			return i, fmt.Errorf("invalid number")
+			return i, ErrInvalidNumber
 		}
 		// Parse the int to check bounds the number of digits could be larger than the max range
 		// We subtract 1 from the index to remove the `i` from our tests
@@ -1417,7 +1433,7 @@ func newFieldsFromBinary(buf []byte) Fields {
 			if valueBuf[0] == '"' {
 				value = unescapeStringField(string(valueBuf[1 : len(valueBuf)-1]))
 				// Check for numeric characters and special NaN or Inf
-			} else if (valueBuf[0] >= '0' && valueBuf[0] <= '9') || valueBuf[0] == '-' || valueBuf[0] == '+' || valueBuf[0] == '.' ||
+			} else if (valueBuf[0] >= '0' && valueBuf[0] <= '9') || valueBuf[0] == '-' || valueBuf[0] == '.' ||
 				valueBuf[0] == 'N' || valueBuf[0] == 'n' || // NaN
 				valueBuf[0] == 'I' || valueBuf[0] == 'i' { // Inf
 

--- a/models/points_test.go
+++ b/models/points_test.go
@@ -362,17 +362,22 @@ func TestParsePointMissingFieldValue(t *testing.T) {
 }
 
 func TestParsePointBadNumber(t *testing.T) {
-	_, err := models.ParsePointsString(`cpu,host=serverA,region=us-west value=1a`)
-	if err == nil {
-		t.Errorf(`ParsePoints("%s") mismatch. got nil, exp error`, `cpu,host=serverA,region=us-west value=1a`)
-	}
-	_, err = models.ParsePointsString(`cpu,host=serverA,region=us-west value=1ii`)
-	if err == nil {
-		t.Errorf(`ParsePoints("%s") mismatch. got nil, exp error`, `cpu,host=serverA,region=us-west value=1ii`)
-	}
-	_, err = models.ParsePointsString(`cpu,host=serverA,region=us-west value=1.0i`)
-	if err == nil {
-		t.Errorf(`ParsePoints("%s") mismatch. got nil, exp error`, `cpu,host=serverA,region=us-west value=1.0i`)
+	for _, tt := range []string{
+		"cpu v=- ",
+		"cpu v=-i ",
+		"cpu v=-. ",
+		"cpu v=. ",
+		"cpu v=1.0i ",
+		"cpu v=1ii ",
+		"cpu v=1a ",
+		"cpu v=-e-e-e ",
+		"cpu v=42+3 ",
+		"cpu v= ",
+	} {
+		_, err := models.ParsePointsString(tt)
+		if err == nil {
+			t.Errorf("Point %q should be invalid", tt)
+		}
 	}
 }
 

